rfcs/0006-merge-operator.md

```diff
@@ -0,0 +1,144 @@
+# SlateDB Merge Operator
+
+Status: Draft
+
+Authors:
+
+* [David Moravek](https://github.com/dmvk)
+
+References:
+
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator
+* https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity
+* https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go
+
+## Motivation
+
+To utilize SlateDB as a fully-featured state backend for a stream processor, it must support three key types of state:
+
+1. **Value**: Latest value or aggregating state.
+2. **List / Bag / Buffering**: State that stores multiple items for later processing.
+3. **Map**: Key-value state for fine-grained data.
+
+State-of-the-art implementations built on RocksDB often leverage the [Merge Operator](https://github.com/facebook/rocksdb/wiki/Merge-Operator) to efficiently manage types 1 and 2. A prime example is list state, where elements are appended continuously until triggered by an event like a window firing. While this method may not always be the absolute optimal solution, it serves as a solid foundation for future iterations and enhancements.
+
+## Goals
+The primary goal is to introduce a new user-facing record type (**Merge Operand**) and a **Merge Operator** interface to SlateDB, allowing applications to bypass the traditional read/modify/update cycle in performance-critical situations where computation can be expressed using an [associative](https://github.com/facebook/rocksdb/wiki/Merge-Operator#associativity-vs-non-associativity) operator.
+
+**Supporting non-associative operators is not a goal of this RFC**. The assumption is that associative operators can cover the majority of use cases while keeping the user-facing API simple and significantly reducing the scope. [Badger](https://github.com/dgraph-io/badger/blob/v4.4.0/merge.go) has adopted the same limitation.
+
+## Public API
+
+We will introduce two new methods for writing Merge Operands into the database. They follow the same structure as existing put methods.
+
+```rust
+impl Db {
+    
+    pub async fn merge(
+        &self,
+        key: &[u8],
+        value: &[u8]
+    ) -> Result<(), SlateDBError> {
+	    ...
+    }
+
+    pub async fn merge_with_options(
+        &self,
+        key: &[u8],
+        value: &[u8],
+        merge_opts: &MergeOptions,
+        write_opts: &WriteOptions,
+    ) -> Result<(), SlateDBError> {
+        ...
+    }
+}
+
+#[derive(Clone, Default)]
+pub struct MergeOptions {
+    pub ttl: Ttl,
```

Comment:
This is a great observation. Using Merge TTL to refresh an already set TTL simply doesn’t work for the reasons you mentioned, unless you’re willing to accept significant complexity and performance costs during compaction.

I did some more research, and there are cases where expiring individual elements in a list is actually desirable (as seen, for example, in Apache Flink). Handling the expiration of individual records separately could be a reasonable limitation for scenarios where Merge with TTLs is needed. We would just need to offer users more granular control over the expiration process 🤔.

1. Merge Operand + Value = Merged Value

Introducing a merged value would allow us to indicate that expiration should be checked with the MergeOperator.

2. Extending the merge operator to define custom TTL logic

This would give users the ability to implement their own expiration logic, ensuring flexibility while using Merge with TTLs. (This is just an outline, we'd have to think bit more about the interface)

```rust
enum ExpiredRecord {
  // Same as a tombstone
  FullyExpired, 
  PartiallyExpired(Bytes, Ttl)
}

trait MergeOperator {

    ...
    
    /// triggers on expiration during compaction/read
    pub fn expire(
        &self,
        value: Bytes,
        i64: ctimestamp,
        ttl: Ttl) -> Return<ExpiredRecord, MergeOperatorError>;
}
```

What do you think?

Reply:
I need to digest this a bit, let's see what other folk think. I generally lean toward reducing API surface area so I'd rather not have another custom expiry mechanism for a pretty niche use case if we can come up with some decent default behavior.

One option is to just not allow TTL on merge values and then just require users to implement their own Compaction filter (#225) once we support that and they can implement TTL however they'd like in userspace.

Reply:
I think that's a fair point. Let's just remove it from MergeOptions for now and properly document that this might misbehave in combination with put & delete TTLs. We can either decide to do this in a followup RFC (I have a feeling that it might be needed at some point) or simply fall back to compaction filters (that's what RocksDB does). This will allow us to reduce surface area of this RFC and move faster.

WDYT?

Reply:
I was thinking about it bit more and It feels like there might also be undefined cases for regular PUT API as well if TTL gets misused 🤔 Let's assume there are two writes to the same key and the earlier one has higher expiration timestamp than the latest one. Is it possible that the DB would actually revert back to the previous value in that case or is there some handling for such cases?

All in all it feels like an API misuse, but I'd love to get some thoughts on how defensive we want to be.

Reply:
I think it's a pretty common use case to have a new PUT "renew" the TTL. In the traditional put case, compaction will replace an expired element with a tombstone with no expiry so you'll never "revive" and older record that hasn't yet expired (see RFC-3).

Anyway, I'm happy not having TTL for merge for now - but let's wait and see what others say.

Reply:
Assuming that we are going to take the route of allowing multiple merge rows for a given key to exist within a single sorted run, then I think one route we could take would be to just not merge rows with different expire times during compaction. Doesn't this then essentially just give you element-level TTL? This does mean that we'd need to differentiate an expired row from a tombstone for merge records though. So when a merge row expires it's recorded as an expired merge record rather than a tombstone, which represents deletion of the whole merge.

Reply:
My 2c:

1. It's most intuitive to me that when I set a TTL for a merge operation, I'm defining the TTL _just_ for that specific merge, not all merges for the key.
2. I'm partial to not supporting TTL for merges for the time being. It sounds like @rodesai's proposal would be additive, so we could support it later if we wanted (provided we change the storage format).
3. If we go @rodesai's route, WDYT about treating regular rows in the same way--two different flags: tombstone and expired. In regular row cases they're the same, but it might allow us to treat merge and regular rows the same in code? Not 100% sure if this is better or worse. But doing a special case flag for merge op feels kinda icky.

Reply:
For the point about two writes to the same key, with earlier write having a longer TTL. In this case, isn't treating the new write's TTL the right thing to do? It is user's intent.

Reply:
Isn't merge just an optimization? It's TTL should behave the same way as a PUT's TTL.

TTL exists at the Db level, and at individual put level. Is this discussion talking about not supporting merge when TTL is at the DB level as well?

Reply:
I've thought about this over the weekend and I'm really starting to like Rohan's and Chris's suggestion.

```
I think one route we could take would be to just not merge rows with different expire times during compaction. Doesn't this then essentially just give you element-level TTL? 
--- 
It's most intuitive to me that when I set a TTL for a merge operation, I'm defining the TTL just for that specific merge, not all merges for the key.
```

This is completely in line with how TTLs for ListState works in Apache Flink for example. It also feels super easy to reason about.

> In the traditional put case, compaction will replace an expired element with a tombstone with no expiry so you'll never "revive" and older record that hasn't yet expired (see RFC-3).

If we go with per-element TTLs this would be different for Merges, we'd simply remove it.

Reply:
> Isn't merge just an optimization? It's TTL should behave the same way as a PUT's TTL.

This is originally what I was thinking as well. It's weird that a merge operator behaves differently from a read-modify-write with the same operator applied in userspace.

The most recent proposal does also make sense to me, particularly for "collect" type merges (e.g. collect a list of elements, and old elements expire as time goes on if there's a TTL to it). I think, however, the behavior is a bit weird for "aggregate" type merge operators. 

Here's an example with a simple "MAX" merge operator with the most recent suggestion:
```
t0: merge(30, ttl-10)
t1: get() -> 30
t5: merge(20, ttl-10)
t12: get() -> 20
t20: get() -> None
```

Hmm... 🤔 

Instead I would imagine the experience should be like a "session aggregate" where we keep the aggregate so long as there hasn't been X time elapsed since the last entry (which might actually be pretty common). That use case would look like:
```
t0: merge(30, ttl-10)
t1: get() -> 30
t5: merge(20, ttl-10)
t12: get() -> 30 // this is what changes because ttl is "renewed"
t20: get() -> None  // enough inactivity, don't care about this key anymore
```

I guess if the user wanted either one behavior or the other they could implement it in userspace without TTL.

This is making me think we should just leave TTL as an extension for merge operators and prohibit it for now.

Reply:
I think there's three scenarios:

1. I want the TTL to renew for the key whenever I insert a new value for the key (that's standard behavior with `put`)
2. I want a TTL per merge element (this is helpful for things like expiring individual elements in a list)
3. I want a key to expire at a specific time, and for the TTL not to renew when I insert a new value for the key.

What is your suggested API for each of these three? If I understand correctly, `ExpireAt` helps for (3), but that seems net new to me (this is cool and I think we should have it). The problem distinguishing between (1) and (2) is still there, we'd need to put that into the codec (which is solvable) but the fundamental problem is that with (1) I can't actually expire anything without knowing it hasn't been updated in a newer SST (that's the original problem I brought up).

Reply:
Right, I think for 1) you simply need to fall back to put. Internal implementation of something like that would most likely need to do READ + UPDATE + PUT under the hood anyway, otherwise we'd need a pretty complex machinery for compactions 🤔

2. and 3. should be covered

It feels like an ok-ish limitation. 1) family of use cases is still supported, but it makes the additional cost explicit to the user.

Reply:
🫡 I'm on board, let's ship it!